### PR TITLE
Building on Windows

### PR DESCRIPTION
**Pull request by crufter**
## Changes

- Added Windows build instructions to the Travis file
- Fixed compile errors
- Fixed tests

## Verification

Ran the test suite.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,16 +3,26 @@ language: go
...
+matrix:
+  include:
+    - os: windows
+      dist: 1803-containers
+      env: VERSION=1.13.2 VARIANT=windows/windowsservercore-1803
+    - os: linux
+      env: VERSION=1.13.2 VARIANT=buster
 
-go_import_path: github.com/improbable-eng/kedge
+go_import_path: github.com/crufter/kedge
```
Comment (ivan-valkov) [inline .travis.yml]: should this be `github.com/improbable-eng/kedge`?

File: pkg/e2e/spinup_test.go [hunk 2]
```diff
@@ -119,7 +119,28 @@ func spinup(t testing.TB, ctx context.Context, cfg config) (chan error, error) {
...
-			err := cmd.Wait()
+			done := make(chan error)
+
+			// Lauching Wait in a goroutine and killing it if it times out.
+			// On windows it seems the executable starts successfully but Wait never returns.
+			go func() { done <- cmd.Wait() }()
+
+			timeout := time.After(5 * time.Second)
+
+			var er error
```
Comment (ivan-valkov) [inline pkg/e2e/spinup_test.go]: nit
```suggestion
			var err error
```
then change the err in the select to something more concrete like `doneErr`

File: pkg/e2e/spinup_test.go [hunk 3]
```diff
@@ -119,7 +119,28 @@ func spinup(t testing.TB, ctx context.Context, cfg config) (chan error, error) {
...
+			select {
+			case <-timeout:
+				// Timeout happened first, kill the process and print a message.
+				cmd.Process.Kill()
```
Comment (ivan-valkov) [inline pkg/e2e/spinup_test.go]: do we need to kill the process here? From what I can see it is killed on line 154.

File: pkg/e2e/spinup_test.go [hunk 4]
```diff
@@ -119,7 +119,28 @@ func spinup(t testing.TB, ctx context.Context, cfg config) (chan error, error) {
...
+				t.Log("Command timed out")
+			case err := <-done:
+				er = err
+				// Command completed before timeout. Print output and error if it exists.
+				t.Logf("Output: %v", stderr.String())
```
Comment (ivan-valkov) [inline pkg/e2e/spinup_test.go]: We already print the stderr on line 146.

File: .travis.yml (hunk 1 above)
Comment (crufter) [reply .travis.yml]: Absolutely right, leftover from fork.

File: pkg/e2e/spinup_test.go (hunk 2 above)
Comment (crufter) [reply pkg/e2e/spinup_test.go]: Agreed :+1:

File: pkg/e2e/spinup_test.go (hunk 3 above)
Comment (crufter) [reply pkg/e2e/spinup_test.go]: @ivan-valkov Yes. The killing at line 154 is in the error callback. The thing is on Windows `cmd.Wait` hangs infinitely, even though the binary starts and stops successfully.

The only thing this change does it kills the binary after 5 seconds, no matter what. Hopefully it won't have any drawbacks, ideally we won't have tests running binaries for that long. Only downside I see is on windows cmd invoking tests will take longer which is okish for now.

Comment (ivan-valkov) [reply pkg/e2e/spinup_test.go]: can we instead add an error to `err` if the timeout expires. Then the error callback will be called and we will do the killing - this way we have this functionality only in 1 place and we are consistent in how we do it.

File: pkg/e2e/spinup_test.go [hunk 5]
```diff
@@ -128,9 +141,9 @@ func spinup(t testing.TB, ctx context.Context, cfg config) (chan error, error) {
 				t.Logf("%s STDOUT\n %s", cmd.Path, stdout.String())
 			}
 
-			return err
+			return doneErr
 		}, func(error) {
-			cmd.Process.Signal(syscall.SIGTERM)
+			cmd.Process.Kill()
```
Comment (crufter) [inline pkg/e2e/spinup_test.go]: For some reason with SIGTERM on Windows Kedge never got interrupted in the e2e tests.

**Comment by crufter** [comment]
I'm also thinking about adding this to the Readme:

```
## Running the test suite on Windows

While on Linux one should be able to run the test suite with `make test`, on Windows some things are not supported out of the box:
- make
- grep

The easiest way to solve this is with [Chocolatey](https://chocolatey.org). You can install both `make` and `grep` with it.
```

File: pkg/e2e/spinup_test.go (hunk 5 above)
Comment (thommay) [reply pkg/e2e/spinup_test.go]: sigterm and sigint aren't _really_ supported by windows; this looks sensible.

**Pull request merged**